File: harness/bundle.py
"""Target bundles: the files that decide how one target behaves, and their hashes.

  python3 harness/bundle.py node:fastify
  python3 harness/bundle.py --all

A run records framework, version, runtime, adapter and serializer. That is everything
about the dependency and nothing about the code that calls it, so when a ratio steps
between two nights the record cannot say whether the wiring changed. A bundle is the
missing half: the target's own wiring, the shared domain module every target in the
language calls, the dependency manifests the version was resolved from, and the
Dockerfiles that pin the runtime. Any of them moves the target, so all of them are hashed.

Two rollups. code_hash covers everything except prose, so correcting a README does not
read as a target that changed. bundle_hash covers prose as well, so a corrected README
does produce a new page. Keeping both is free once files are listed individually, and
conflating them would make one of the two behaviours wrong.

The rules are versioned as bundle-v1, because changing any of them silently reissues
every hash in the series:

  - the file set comes from git ls-files, never a filesystem walk, so an editor backup or
    a stale build artifact cannot enter it
  - paths are repo-relative with forward slashes, sorted bytewise
  - each file is SHA-256 over its raw bytes on disk, full hex digest
  - .gitattributes pins * text=auto eol=lf, without which a Windows checkout changes
    every hash
  - the rollup is SHA-256 over the canonical manifest text, not over a JSON serialisation

Bytes come from the working tree rather than from the commit, because the hash has to
describe what actually ran. That makes a dirty tree detectable: the recorded hash will
not verify against `git show <commit>:<path>`, and docs/bundles.html §4 has the site say
the source is unavailable rather than render today's file under an old number.

Still to build: the framework metadata block in docs/bundles.html §5, which derives repo,
licence and homepage from the resolved dependency tree. It needs no run to produce it, so
unlike the hashes it can be added to the series at any time.
"""
import argparse, functools, hashlib, json, os, pathlib, re, subprocess, sys

ROOT = pathlib.Path(__file__).resolve().parent.parent
SPEC = ROOT / "spec"
MATRIX = json.loads((SPEC / "matrix.json").read_text())

BUNDLE_VERSION = "bundle-v1"


# The two directories every target in a language shares: the domain module they all call,
# and the host entry points that invoke them. Java spells both without the underscore.
SHARED_DIR = {"java": "shared"}
HOST_DIR = {"java": "hosts"}

MANIFEST_NAMES = {"package.json", "package-lock.json", "go.mod", "go.sum", "pom.xml",
                  "Cargo.toml", "Cargo.lock"}
CONFIG_SUFFIXES = (".properties", ".yaml", ".yml", ".toml", ".ini", ".conf")

# Host entry points that sit inside a target's own directory rather than in the shared
# host module: Go names the file, Java gives it a package.
HOST_IN_TARGET = re.compile(r"/rb/(lambda|gcp)/|/lambda\.go$")

# A bundle missing any of these did not resolve. See check().
REQUIRED_ROLES = ("source", "shared", "manifest")


def target_dir(name):
    """A target's directory is its name. Nothing is special-cased any more."""
    return name


def shared_dir(language):
    return SHARED_DIR.get(language, "_shared")


def host_dir(language):
    return HOST_DIR.get(language, "_hosts")


def git(*args):
    out = subprocess.run(["git", *args], cwd=ROOT, capture_output=True, text=True)
    if out.returncode != 0:
        raise RuntimeError("git %s: %s" % (" ".join(args), out.stderr.strip()))
    return out.stdout


def git_bytes(*args):
    """File contents are hashed, so they have to arrive as bytes and never through a
    decode. A source file is not always valid UTF-8 and a round trip would not be identity
    even when it is."""
    out = subprocess.run(["git", *args], cwd=ROOT, capture_output=True)
    if out.returncode != 0:
        raise RuntimeError("git %s: %s" % (" ".join(args), out.stderr.decode().strip()))
    return out.stdout


# Every lookup below takes `at`: None reads the working tree, a commit reads history.
# A run records the commit it measured, so the site rebuilds that run's manifest from
# history rather than from today's files, which is the whole point of recording it.
@functools.lru_cache(maxsize=None)
def tracked(prefix, at=None):
    if at is None:
        return tuple(p for p in git("ls-files", "-z", "--", prefix).split("\0") if p)
    return tuple(p for p in git("ls-tree", "-r", "-z", "--name-only", at, "--", prefix)
                 .split("\0") if p)


@functools.lru_cache(maxsize=None)
def matrix_at(at):
    """bundle_roots as it stood at that commit. Reading today's would misresolve a target
    whose roots were declared later, and the rollup would then fail to verify against a
    perfectly good run."""
    return MATRIX if at is None else json.loads(git("show", "%s:spec/matrix.json" % at))


def blob(path, at=None):
    if at is None:
        return (ROOT / path).read_bytes()
    return git_bytes("cat-file", "blob", "%s:%s" % (at, path))


def roots(language, target, at=None):
    """The paths a target's file set is drawn from.

    A path ending in / contributes only the files sitting directly in it. Any other path
    contributes its whole subtree.

    A target the convention does not fit declares its own list under bundle_roots in
    spec/matrix.json, keyed language:target, the way the host exclusions are declared.
    """
    declared = matrix_at(at).get("bundle_roots", {}).get("%s:%s" % (language, target))
    if declared:
        return list(declared)
    lang = "targets/" + language
    return ["%s/%s" % (lang, target_dir(target)),
            "%s/%s" % (lang, shared_dir(language)),
            "%s/%s" % (lang, host_dir(language)),
            lang + "/"]


def files(language, target, at=None):
    """Every tracked file in the bundle, sorted.

    The language directory contributes only the manifests and Dockerfiles sitting directly
    in it. Listing it whole would sweep in every sibling target.
    """
    wanted = set()
    for root in roots(language, target, at):
        if root.endswith("/"):
            wanted.update(p for p in tracked(root.rstrip("/"), at)
                          if "/" not in p[len(root):])
        else:
            wanted.update(tracked(root, at))
    if not wanted:
        raise RuntimeError("%s:%s has no tracked files under %s"
                           % (language, target, ", ".join(roots(language, target, at))))
    # The rule is bytewise, which for UTF-8 is what sorting the strings already does:
    # byte order and code point order agree. Only something locale-aware would differ,
    # and that would differ between machines.
    return sorted(wanted)


def role(language, path):
    name = path.rsplit("/", 1)[-1]
    if name.endswith(".md"):
        return "prose"
    if name in MANIFEST_NAMES:
        return "manifest"
    if name.endswith(CONFIG_SUFFIXES):
        return "config"
    if (name.startswith("Dockerfile") or "/%s/" % host_dir(language) in path
            or HOST_IN_TARGET.search(path)):
        return "host"
    if "/%s/" % shared_dir(language) in path:
        return "shared"
    return "source"


def sha256(path):
    h = hashlib.sha256()
    with open(path, "rb") as fh:
        for block in iter(lambda: fh.read(1 << 20), b""):
            h.update(block)
    return h.hexdigest()


def rollup(entries):
    """SHA-256 over one '<hash>  <path>' line per file, in sorted order.

    Hashing a JSON serialisation would make the digest depend on key order, spacing and
    how a language's encoder escapes; a line-oriented form has one representation. It is
    sha256sum's own format, so the canonical text verifies with `sha256sum -c` from the
    repository root.
    """
    text = "".join("%s  %s\n" % (e["hash"][7:], e["path"]) for e in entries)
    return "sha256:" + hashlib.sha256(text.encode()).hexdigest()


def commit():
    return git("rev-parse", "HEAD").strip()


def repo():
    """The owner/name slug. Recorded rather than hard-coded so a fork's pages link to the
    fork's own code."""
    slug = os.environ.get("GITHUB_REPOSITORY", "")
    if slug:
        return slug
    m = re.search(r"[:/]([^/:]+/[^/]+?)(?:\.git)?$", git("remote", "get-url", "origin").strip())
    return m.group(1) if m else ""


@functools.lru_cache(maxsize=None)
def pushed(at):
    """Whether a remote-tracking branch holds this commit.

    A permalink to a commit that was never pushed is a 404, which reads as the code having
    been deleted rather than as the run having been local. Nightly runs are Actions runs on
    a pushed commit, so this only ever hides a link for a local run.
    """
    try:
        return bool(git("branch", "-r", "--contains", at).strip())
    except RuntimeError:
        return False


def manifest(language, target, at=None):
    entries = []
    for path in files(language, target, at):
        raw = blob(path, at)
        entries.append({"path": path, "role": role(language, path),
                        "bytes": len(raw),
                        "hash": "sha256:" + hashlib.sha256(raw).hexdigest()})
    return {"bundle_version": BUNDLE_VERSION, "target": "%s:%s" % (language, target),
            "bundle_hash": rollup(entries),
            "code_hash": rollup([e for e in entries if e["role"] != "prose"]),
            "commit": at or commit(), "files": entries}


def hashes(language, target):
    """Just the two rollups, which is all a run record carries."""
    m = manifest(language, target)
    return {"bundle_hash": m["bundle_hash"], "code_hash": m["code_hash"]}


def check(pairs):
    """Return one complaint per target whose bundle did not resolve.

    A target with no wiring of its own, no shared domain or no dependency manifest is the
    convention in roots() failing to fit a language and producing a partial file set. A
    partial set hashes as cleanly as a complete one, so nothing else would notice.
    """
    bad = []
    for language, name in pairs:
        try:
            roles = {e["role"] for e in manifest(language, name)["files"]}
        except Exception as e:
            bad.append("%s:%s did not resolve (%s)" % (language, name, e))
            continue
        missing = [r for r in REQUIRED_ROLES if r not in roles]
        if missing:
            bad.append("%s:%s has no %s file" % (language, name, " and no ".join(missing)))
    return bad


def implemented():
    for language, spec in MATRIX["languages"].items():
        for name in spec["implemented"]:
            yield language, name


def main():
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("targets", nargs="*", metavar="language:target")
    ap.add_argument("--all", action="store_true", help="every implemented target")
    ap.add_argument("--summary", action="store_true",
                    help="one line per target instead of the full manifest")
    ap.add_argument("--check", action="store_true",
                    help="assert every named bundle resolved; print nothing else")
    a = ap.parse_args()

    pairs = list(implemented()) if a.all else []
    for entry in a.targets:
        language, _, name = entry.partition(":")
        if not name:
            sys.exit("target %r has no language: write language:target" % entry)
        if language not in MATRIX["languages"]:
            sys.exit("unknown language %r in target %r" % (language, entry))
        pairs.append((language, name))
    if not pairs:
        sys.exit("no targets: name some, or pass --all")

    if a.check:
        bad = check(pairs)
        for line in bad:
            print(line)
        print("%d/%d bundles resolve" % (len(pairs) - len(bad), len(pairs)))
        return 1 if bad else 0

    for language, name in pairs:
        m = manifest(language, name)
        if a.summary:
            print("%-22s code %s  bundle %s  %2d files"
                  % (m["target"], m["code_hash"][7:19], m["bundle_hash"][7:19],
                     len(m["files"])))
        else:
            print(json.dumps(m, indent=2))


if __name__ == "__main__":
    sys.exit(main() or 0)

File: orchestrator/bundle.ts
// Bundles: every tracked file that decides how a framework, or the tests, behave, each with a
// role and a hash, rolled up into codeHash and bundleHash.
//
// A run records the two rollups and the commit. The files themselves are never stored: the site
// rebuilds a bundle from git at the recorded commit and refuses to show code whose rollup does
// not verify, because rendering today's file under an old number is the mistake this exists to
// prevent. Upstream's harness/bundle.py and docs/bundles.html are the design.
//
// Two rollups, because a corrected README or a sharpened test should produce a new page without
// reading as a framework that changed. codeHash leaves out the roles that do not change what
// runs; bundleHash covers every file.
//
// The rules below are bundle-v1. Changing any of them reissues every hash in the series, so a
// change is a new version:
//   - the file set comes from git, never a filesystem walk, so an editor backup or a build
//     artifact cannot enter it
//   - a framework's bundle is read for one host, and leaves out the directories named for the
//     other hosts, so a change to how one host starts the framework is no change on another
//   - paths are repo-relative and sorted bytewise
//   - a file's hash is SHA-256 over its bytes, as sha256:<hex>
//   - a rollup is SHA-256 over one `<hex>  <path>` line per file, which is sha256sum's own
//     format, so `sha256sum -c` verifies it from the repository root
//
// Without `at`, bytes come from the working tree, because a hash has to describe what actually
// ran. A dirty tree is then visible: its rollup does not verify against the commit.
import { createHash } from "node:crypto";

import { blob, resolveCommit, tracked } from "./git.ts";
import { HOST_IDS, type HostId } from "./hosts.ts";

export const BUNDLE_VERSION = "bundle-v1";

export type FrameworkRole = "source" | "manifest" | "config" | "host" | "contract" | "prose" | "test" | "client";
export type TestsRole = "test" | "family" | "kit" | "model" | "payload" | "snapshot" | "manifest" | "source" | "prose";
export type Role = FrameworkRole | TestsRole;

export interface BundleFile {
  /** Repo-relative, forward slashes. */
  readonly path: string;
  readonly role: Role;
  readonly bytes: number;
  readonly hash: string;
}

export interface Bundle {
  readonly bundleVersion: typeof BUNDLE_VERSION;
  /** `<language>:<name>` for a framework, `tests` for the tests. */
  readonly id: string;
  /** The commit the files were read at, or HEAD's when they came from the working tree. */
  readonly commit: string;
  readonly bundleHash: string;
  readonly codeHash: string;
  readonly files: readonly BundleFile[];
}

const sha256 = (bytes: Buffer | string): string => createHash("sha256").update(bytes).digest("hex");

/** SHA-256 over one `<hex>  <path>` line per file, in the order given, which is sorted. */
export function rollup(files: readonly Pick<BundleFile, "path" | "hash">[]): string {
  return `sha256:${sha256(files.map((f) => `${f.hash.slice("sha256:".length)}  ${f.path}\n`).join(""))}`;
}

function build(
  root: string,
  id: string,
  prefix: string,
  at: string | undefined,
  roleOf: (path: string) => Role,
  notCode: ReadonlySet<Role>,
  keep: (path: string) => boolean = () => true,
): Bundle {
  const paths = tracked(root, prefix, at).filter(keep);
  if (paths.length === 0) throw new Error(`${id} has no tracked files under ${prefix}`);
  const files = paths.map((path): BundleFile => {
    const bytes = blob(root, path, at);
    return { path, role: roleOf(path), bytes: bytes.length, hash: `sha256:${sha256(bytes)}` };
  });
  return {
    bundleVersion: BUNDLE_VERSION,
    id,
    commit: resolveCommit(root, at),
    bundleHash: rollup(files),
    codeHash: rollup(files.filter((f) => !notCode.has(f.role))),
    files,
  };
}

/**
 * The rollup of every tracked file under `prefix`, for something whose identity is its files
 * and which has no roles to tell apart, such as the traffic generator.
 */
export function treeRollup(root: string, prefix: string, at?: string): string {
  return rollup(tracked(root, prefix, at).map((path) => ({ path, hash: `sha256:${sha256(blob(root, path, at))}` })));
}

// ---- frameworks ------------------------------------------------------------------------

export interface FrameworkKey {
  readonly language: string;
  readonly name: string;
}

export const frameworkDir = (f: FrameworkKey): string => `frameworks/${f.language}/${f.name}`;

/** The name a reader types and a run records: `dotnet:carter`. */
export const frameworkId = (f: FrameworkKey): string => `${f.language}:${f.name}`;

const MANIFEST_NAMES = new Set([
  "rb.json",
  "package.json",
  "package-lock.json",
  "pnpm-lock.yaml",
  "yarn.lock",
  "go.mod",
  "go.sum",
  "pom.xml",
  "Cargo.toml",
  "Cargo.lock",
  "requirements.in",
  "requirements.txt",
  "pyproject.toml",
  "uv.lock",
  "Directory.Packages.props",
  "Directory.Build.props",
  "packages.lock.json",
  "nuget.config",
  "global.json",
]);
/** A .NET project or solution is named after itself, so it is matched by suffix. */
const MANIFEST_SUFFIXES = [".csproj", ".sln", ".slnx"];
const CONFIG_SUFFIXES = [".properties", ".yaml", ".yml", ".toml", ".ini", ".conf"];
/**
 * An API description a framework routes from says what it serves, so it is code. Matched by
 * name, because a .yaml beside a framework is far more often a setting.
 */
const CONTRACT_NAMES = new Set(["openapi.yaml", "openapi.yml", "openapi.json"]);
/**
 * The TypeScript that reads this framework's error bodies. The benchmark holds the framework to
 * it and the framework never runs it, so it is in the bundle as a test and out of codeHash.
 */
const CLIENT_EXCEPTION = "client-exception/";
/**
 * The OpenAPI document the framework writes about itself and the client generated from it. The
 * image never reads them, so they stay out of codeHash, and the handler finder never reads them,
 * because every route literal is in the document a second time.
 */
const CLIENT = "Client/";
/** Go keeps a test beside the code it tests, so it is told apart by name, as `go build` does. */
const TEST_SUFFIXES = ["_test.go"];
const FRAMEWORK_NOT_CODE: ReadonlySet<Role> = new Set(["prose", "test", "client"]);

/** A bundle missing any of these did not resolve, and a partial file set hashes as cleanly as a whole one. */
const FRAMEWORK_REQUIRED: readonly Role[] = ["source", "manifest", "host"];

const inside = (rel: string, dir: string): boolean => {
  const d = dir.replace(/^\.\/|\/$/g, "");
  return d === "" || d === "." || rel === d || rel.startsWith(`${d}/`);
};

/** The role of a file, from its path relative to the framework's directory. */
export function frameworkRole(rel: string, suitePaths: readonly string[]): FrameworkRole {
  const name = rel.slice(rel.lastIndexOf("/") + 1);
  // Before the name rules: a suite's own package.json is not a dependency of the framework.
  if (rel.startsWith(CLIENT_EXCEPTION) || suitePaths.some((p) => inside(rel, p)) || TEST_SUFFIXES.some((s) => name.endsWith(s))) {
    return "test";
  }
  // Before the name rules: Client/openapi.json is what the framework says it serves, not a contract it routes from.
  if (rel.startsWith(CLIENT)) return "client";
  if (CONTRACT_NAMES.has(name)) return "contract";
  if (name.endsWith(".md")) return "prose";
  if (MANIFEST_NAMES.has(name) || MANIFEST_SUFFIXES.some((s) => name.endsWith(s))) return "manifest";
  if (CONFIG_SUFFIXES.some((s) => name.endsWith(s))) return "config";
  if (name.startsWith("Dockerfile") || name === ".dockerignore") return "host";
  return "source";
}

/**
 * The suite's directories, as rb.json named them at the same commit. Read here rather than
 * from the loaded manifest so a bundle at an old commit uses the roles that commit declared.
 * Lenient on purpose: whether rb.json is valid is the loader's question, and a bundle has to
 * resolve for a framework whose manifest is wrong so the site can still say what it was.
 */
function suitePathsAt(root: string, dir: string, at: string | undefined): string[] {
  let raw: unknown;
  try {
    raw = JSON.parse(blob(root, `${dir}/rb.json`, at).toString("utf8"));
  } catch {
    return [];
  }
  const paths = (raw as { suite?: { paths?: unknown } } | null)?.suite?.paths;
  return Array.isArray(paths) ? paths.filter((p): p is string => typeof p === "string") : [];
}

/** The directories of every host but this one, relative to the framework's directory. */
const otherHosts = (host: HostId): string[] => HOST_IDS.filter((h) => h !== host).map((h) => `${h}/`);

export function frameworkBundle(root: string, f: FrameworkKey, host: HostId, at?: string): Bundle {
  const dir = frameworkDir(f);
  const suite = suitePathsAt(root, dir, at);
  const others = otherHosts(host);
  return build(
    root,
    frameworkId(f),
    `${dir}/`,
    at,
    (path) => frameworkRole(path.slice(dir.length + 1), suite),
    FRAMEWORK_NOT_CODE,
    (path) => !others.some((o) => path.slice(dir.length + 1).startsWith(o)),
  );
}

/** One complaint per role a framework bundle should hold and does not. */
export function frameworkProblems(b: Bundle): string[] {
  const roles = new Set(b.files.map((f) => f.role));
  const missing = FRAMEWORK_REQUIRED.filter((r) => !roles.has(r));
  return missing.length === 0 ? [] : [`${b.id} has no ${missing.join(" and no ")} file`];
}

// ---- the tests ---------------------------------------------------------------------------

/** The tests are a bundle of their own, so the site can show each test's source beside its numbers. */
export const TESTS_ID = "tests";
const TESTS_DIR = "tests";
const RESERVED = new Set(["kit", "models", "payloads"]);
/**
 * Snapshots hold the error bodies frameworks were captured writing, which only the reference
 * in orchestrator/test reads. Changing one changes no request and no answer a run checks.
 * Which questions were asked is the corpus version's business, not codeHash's: prose inside a
 * test file moves codeHash and leaves the corpus version alone.
 */
const TESTS_NOT_CODE: ReadonlySet<Role> = new Set(["prose", "snapshot"]);

export function testsRole(path: string): TestsRole {
  const seg = path.split("/").slice(1);
  const name = seg[seg.length - 1]!;
  if (name.endsWith(".md")) return "prose";
  if (name.endsWith(".snap.json")) return "snapshot";
  if (seg.length === 1) return name === "package.json" ? "manifest" : name === "index.ts" ? "family" : "source";
  const top = seg[0]!;
  if (top === "kit") return "kit";
  if (top === "models") return "model";
  if (top === "payloads") return "payload";
  if (seg.length === 2 && name.endsWith(".ts")) return name === "index.ts" ? "family" : "test";
  return "source";
}

export function testsBundle(root: string, at?: string): Bundle {
  return build(root, TESTS_ID, `${TESTS_DIR}/`, at, testsRole, TESTS_NOT_CODE);
}

/** The test file each corpus id is declared in, from the bundle alone: `json.small` is `tests/json/small.ts`. */
export function testFiles(b: Bundle): Map<string, BundleFile> {
  const out = new Map<string, BundleFile>();
  for (const f of b.files) {
    if (f.role !== "test") continue;
    const [, family, file] = f.path.split("/");
    if (family === undefined || file === undefined || RESERVED.has(family)) continue;
    // The file name is the id's name with dashes where the id has underscores.
    out.set(`${family}.${file.slice(0, -".ts".length).replaceAll("-", "_")}`, f);
  }
  return out;
}
